pkg/kconfig/kconfig.go

```diff
@@ -0,0 +1,190 @@
+var (
+	configPaths = []string{
+		"/proc/config.gz",
+		"/boot/config",
+		"/boot/config-%s",
+	}
+)
+
+var ebpfCheckOptions = []string{
+	"CONFIG_BPF",
+	"CONFIG_BPF_SYSCALL",
+	"CONFIG_HAVE_EBPF_JIT",
+	"CONFIG_BPF_JIT",
+	"CONFIG_BPF_JIT_ALWAYS_ON",
+	"CONFIG_BPF_EVENTS",
+}
+
+func IsBPFEnabled() (bool, error) {
+	var e error
+
+	KernelConfig, e := getConfig()
+	if e != nil {
+		return false, e
+	}
+
+	for _, option := range ebpfCheckOptions {
+		value, found := KernelConfig[option]
+		if !found {
+			return false, fmt.Errorf("kernel Config required for ebpf not found, Config Option:%s", option)
+		}
+
+		if value != "y" {
+			return false, fmt.Errorf("kernel Config required for ebpf is disabled, Config Option:%s", option)
+		}
+	}
+	return true, nil
+}
+
+func IsStackUnwindingEnabled() (bool, error) {
+	var e error
+
+	KernelConfig, e := getConfig()
+	if e != nil {
+		return false, e
+	}
+
+	value, found := KernelConfig[StackUnwinder]
+	if !found {
+		return false, fmt.Errorf("kernel config required for frame pointer stack unwinder not found, Config Option:%s", StackUnwinder)
+	}
+
+	if value != "y" {
+		return false, fmt.Errorf("kernel config for frame pointer stack unwinder is disabled, Config Option:%s", StackUnwinder)
+	}
+	return true, nil
+}
+
+func getConfig() (map[string]string, error) {
+	var found bool
+	KernelConfig := make(map[string]string)
+
+	i, e := UnameRelease()
+	if e != nil {
+		return KernelConfig, e
+	}
+
+	for _, configPath := range configPaths {
+		if strings.Contains(configPath, ".gz") {
+			KernelConfig, e = readConfigFromProc(configPath)
+		}
+
+		if strings.Contains(configPath, "%s") {
+			bootConf := fmt.Sprintf(configPath, i)
+			KernelConfig, e = readConfigFromBoot(bootConf)
+		}
+
+		if e != nil {
+			continue
+		}
+
+		if len(KernelConfig) > 0 {
+			found = true
+		}
+	}
+
+	if !found {
+		return nil, fmt.Errorf("kernel config not found")
+	}
+	return KernelConfig, nil
+}
+
+func readConfigFromBoot(filename string) (map[string]string, error) {
+	KernelConfig := make(map[string]string)
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return KernelConfig, err
+	}
+	defer f.Close()
+
+	s := bufio.NewScanner(f)
+	if err = parse(s, KernelConfig); err != nil {
+		return KernelConfig, err
+	}
+	return KernelConfig, nil
+}
+
+func readConfigFromProc(filename string) (map[string]string, error) {
+	KernelConfig := make(map[string]string)
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return KernelConfig, err
+	}
+	defer f.Close()
+
+	zreader, err := gzip.NewReader(f)
+	if err != nil {
+		return KernelConfig, err
+	}
+	defer zreader.Close()
+
+	s := bufio.NewScanner(zreader)
+	if err = parse(s, KernelConfig); err != nil {
+		return KernelConfig, err
+	}
+	return KernelConfig, nil
+}
+
+// Returns true is the process is running in a container
+// TODO: Add a container detection via Sched to cover more scenarios
+// https://man7.org/linux/man-pages/man7/sched.7.html
+func IsInContainer() (bool, error) {
```

Comment:
I don't know if it's possible, but we need to check if it's running in privileged mode. Otherwise, the info about running in a container is not actionable.

Reply:
Yeah, AFAIU it's not easy to check if it's running in a privileged mode. I'll check via sched though. Another idea was that we can try running some command(e.g. iproute2) that's only allowed in the privileged mode to detect that. But I'm not sure if that's worth it as everything is warning at the moment. Although I think it's still a useful Log for the user when the kernel config read fails because they can at least know that this could be a possible reason.